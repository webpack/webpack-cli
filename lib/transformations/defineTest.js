'use strict';

const fs = require('fs');
const path = require('path');

/**
 * Utility function to run a jscodeshift script within a unit test. This makes
 * several assumptions about the environment:
 *
 * - `dirName` contains the name of the directory the test is located in. This
 *   should normally be passed via __dirname.
 * - The test should be located in a subdirectory next to the transform itself.
 *   Commonly tests are located in a directory called __tests__.
 * - `transformName` contains the filename of the transform being tested,
 *   excluding the .js extension.
 * - `testFilePrefix` optionally contains the name of the file with the test
 *   data. If not specified, it defaults to the same value as `transformName`.
 *   This will be suffixed with ".input.js" for the input file and ".output.js"
 *   for the expected output. For example, if set to "foo", we will read the
 *   "foo.input.js" file, pass this to the transform, and expect its output to
 *   be equal to the contents of "foo.output.js".
 * - Test data should be located in a directory called __testfixtures__
 *   alongside the transform and __tests__ directory.
 */
function runSingleTansform(dirName, transformName, testFilePrefix) {
	if (!testFilePrefix) {
		testFilePrefix = transformName;
	}

	const fixtureDir = path.join(dirName, '__testfixtures__');
	const inputPath = path.join(fixtureDir, testFilePrefix + '.input.js');
	const source = fs.readFileSync(inputPath, 'utf8');
	// Assumes transform and test are on the same level
	const module = require(path.join(dirName, transformName + '.js'));
	// Handle ES6 modules using default export for the transform
	const transform = module.default ? module.default : module;

	// Jest resets the module registry after each test, so we need to always get
	// a fresh copy of jscodeshift on every test run.
	let jscodeshift = require('jscodeshift/dist/core');
	if (module.parser) {
		jscodeshift = jscodeshift.withParser(module.parser);
	}
	const ast = jscodeshift(source);
	return transform(jscodeshift, ast, source).toSource({ quote: 'single' });
}

/**
 * Handles some boilerplate around defining a simple jest/Jasmine test for a
 * jscodeshift transform.
 */
function defineTest(dirName, transformName, testFilePrefix) {
	const testName = testFilePrefix
		? `transforms correctly using "${testFilePrefix}" data`
		: 'transforms correctly';
	describe(transformName, () => {
		it(testName, () => {
			const output = runSingleTansform(dirName, transformName, testFilePrefix);
			expect(output).toMatchSnapshot();
		});
	});
}
module.exports = defineTest;
