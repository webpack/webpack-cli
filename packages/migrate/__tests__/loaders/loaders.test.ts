import defineTest from "../../../utils/__tests__/defineTest";
import { join } from "path";

const dirName: string = join(__dirname);

describe('loaders', function() {
    {
        defineTest(dirName, "loaders", "loaders-0");
        defineTest(dirName, "loaders", "loaders-1");
        defineTest(dirName, "loaders", "loaders-2");
        defineTest(dirName, "loaders", "loaders-3");
        defineTest(dirName, "loaders", "loaders-4");
        defineTest(dirName, "loaders", "loaders-5");
        defineTest(dirName, "loaders", "loaders-6");
        defineTest(dirName, "loaders", "loaders-7");
        defineTest(dirName, "loaders", "loaders-8");
    }
});


