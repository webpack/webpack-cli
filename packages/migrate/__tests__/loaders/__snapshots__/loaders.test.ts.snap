// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loaders loaders transforms correctly using "loaders-0" data 1`] = `
"module.exports = [
  {
	module: {
		rules: [
		{
			use: [{
                loader: 'babel-loader'
            }],
			test: /\\\\.js$/,
		},
		],
	},
  },
  {
	module: {
		rules: [
		{
			use: [{
                loader: 'style-loader'
            }, {
                loader: 'css-loader',

                options: {
                    modules: true,
                    importLoaders: 1,
                    string: 'test123'
                }
            }],
			test: /\\\\.css$/,
		},
		],
	},
  },
  {
	module: {
		rules: [
		{
			use: [{
				loader: 'style-loader',
			}, {
				loader: 'css-loader',
				options: {
				modules: true,
				},
			}],
			test: /\\\\.css$/,
		},
		],
	},
  },
  {
	module: {
		rules: [
		{
            use: [{
                loader: 'eslint-loader'
            }],

            test: /\\\\.js$/,
            enforce: 'pre'
        },
		],
	},
  },
  {
	module: {
		rules: [
		{
            use: [{
                loader: 'my-post-loader'
            }],

            test: /\\\\.js$/,
            enforce: 'post'
        },
		],
	},
  },
  {
	module: {
        rules: [{
			use: [{
                loader: 'babel-loader'
            }],
			test: /\\\\.js$/,
		}, {
            use: [{
                loader: 'eslint-loader'
            }],

            test: /\\\\.js$/,
            enforce: 'pre'
        }]
    },
  },
  {
	module: {
        rules: [{
			use: [{
                loader: 'babel-loader'
            }],
			test: /\\\\.js$/,
		}, {
            use: [{
                loader: 'my-post-loader'
            }],

            test: /\\\\.js$/,
            enforce: 'post'
        }]
    },
  },
];
"
`;

exports[`loaders loaders transforms correctly using "loaders-1" data 1`] = `
"module.exports = {
  module: {
	rules: [
		{
		use: [{
          loader: 'style-loader'
        }, {
          loader: 'css-loader',

          options: {
            modules: true,
            importLoaders: 1,
            string: 'test123'
          }
        }],
		test: /\\\\.css$/,
		},
	],
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-2" data 1`] = `
"module.exports = {
  module: {
	rules: [
		{
		use: [{
        loader: 'style-loader',
        }, {
        loader: 'css-loader',
        options: {
            modules: true,
        },
        }],
		test: /\\\\.css$/,
		},
	],
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-3" data 1`] = `
"module.exports = {
  module: {
	rules: [
		{
          use: [{
            loader: 'eslint-loader'
          }],

          test: /\\\\.js$/,
          enforce: 'pre'
        },
	],
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-4" data 1`] = `
"module.exports = {
  module: {
	rules: [
		{
          use: [{
            loader: 'my-post-loader'
          }],

          test: /\\\\.js$/,
          enforce: 'post'
        },
	],
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-5" data 1`] = `
"module.exports = {
  module: {
    rules: [{
    use: [{
      loader: 'babel-loader'
    }],
    test: /\\\\.js$/,
    }, {
      use: [{
        loader: 'eslint-loader'
      }],

      test: /\\\\.js$/,
      enforce: 'pre'
    }]
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-6" data 1`] = `
"module.exports = {
  module: {
    rules: [{
    use: [{
      loader: 'babel-loader'
    }],
    test: /\\\\.js$/,
    }, {
      use: [{
        loader: 'my-post-loader'
      }],

      test: /\\\\.js$/,
      enforce: 'post'
    }]
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-7" data 1`] = `
"module.exports = {
  module: {
	rules: [
		{
          exclude: /(node_modules)/,

          use: [{
            loader: 'babel-loader',

            options: {
                presets: [\\"@babel/preset-env\\"],
            }
          }],

          test: /\\\\.js$/
        },
	],
  },
};
"
`;

exports[`loaders loaders transforms correctly using "loaders-8" data 1`] = `
"module.exports = {
	module: {
		rules: [
			{
				include: path.join(__dirname, \\"src\\"),
				use: [{
                    loader: 'style-loader'
                }, {
                    loader: 'css-loader',

                    options: {
                        modules: true,
                        importLoaders: 1,
                        string: 'test123'
                    }
                }],
				test: /\\\\.js$/
			}
		]
	}
};
"
`;
